File: packages/drive-detector/ios/Sources/DriveDetectorCore/Distance.swift
import Foundation

public enum Distance {
    /// Polomer Zeme v metroch (stredný).
    static let earthRadius: Double = 6_371_000

    /// Haversine. CLLocation.distance(from:) by bol presnejší, ale ten sem
    /// nesmie — jadro musí ísť otestovať bez CoreLocation.
    public static func meters(from a: TripPoint, to b: TripPoint) -> Double {
        let toRad = Double.pi / 180
        let dLat = (b.lat - a.lat) * toRad
        let dLng = (b.lng - a.lng) * toRad
        let lat1 = a.lat * toRad
        let lat2 = b.lat * toRad
        let h = pow(sin(dLat / 2), 2) + pow(sin(dLng / 2), 2) * cos(lat1) * cos(lat2)
        return 2 * earthRadius * asin(min(1, sqrt(h)))
    }

    /// Súčet vzdialeností medzi po sebe idúcimi bodmi.
    public static func total(of points: [TripPoint]) -> Double {
        guard points.count > 1 else { return 0 }
        var spolu: Double = 0
        for i in 1..<points.count {
            spolu += meters(from: points[i - 1], to: points[i])
        }
        return spolu
    }

    public static func maxSpeed(of points: [TripPoint]) -> Double {
        points.reduce(0) { max($0, $1.speedKmh) }
    }
}

File: packages/drive-detector/android/src/main/java/sk/faktero/drivedetector/core/Distance.java
package sk.faktero.drivedetector.core;

import java.util.List;

public final class Distance {
    /** Polomer Zeme v metroch (stredný). */
    private static final double POLOMER_ZEME = 6_371_000d;

    private Distance() {}

    /**
     * Haversine. `Location.distanceBetween` by bol presnejší, ale ten sem
     * nesmie — jadro musí ísť otestovať bez tried Androidu.
     */
    public static double meters(TripPoint a, TripPoint b) {
        double toRad = Math.PI / 180d;
        double dLat = (b.lat - a.lat) * toRad;
        double dLng = (b.lng - a.lng) * toRad;
        double lat1 = a.lat * toRad;
        double lat2 = b.lat * toRad;
        double h = Math.pow(Math.sin(dLat / 2), 2)
                + Math.pow(Math.sin(dLng / 2), 2) * Math.cos(lat1) * Math.cos(lat2);
        return 2 * POLOMER_ZEME * Math.asin(Math.min(1, Math.sqrt(h)));
    }

    /** Súčet vzdialeností medzi po sebe idúcimi bodmi. */
    public static double total(List<TripPoint> points) {
        if (points == null || points.size() < 2) return 0;
        double spolu = 0;
        for (int i = 1; i < points.size(); i++) spolu += meters(points.get(i - 1), points.get(i));
        return spolu;
    }

    public static double maxSpeed(List<TripPoint> points) {
        double max = 0;
        if (points == null) return 0;
        for (TripPoint p : points) max = Math.max(max, p.speedKmh);
        return max;
    }
}
